File: python-algorithm/ds_1_stack_queue/L232.cpp
//
// Created by 李恒 on 2020/7/2.
//
#include <iostream>
#include <stack>
using  namespace std;

class MyQueue {
private:
    stack<int> s1;
    stack<int> s2;
public:
    /** Initialize your data structure here. */
    MyQueue() {

    }

    /** Push element x to the back of queue. */
    void push(int x) {
        s1.push(x);
    }

    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        if(s1.empty() && s2.empty()) //可以拿掉，题目条件:空队列不会调用pop peek
            throw runtime_error("Queue is empty!") ;
        if(s2.empty()){
            while(!s1.empty()){
                s2.push(s1.top());
                s1.pop() ;
            }
        }
        int re = s2.top() ;
        s2.pop() ;
        return re ;
    }

    /** Get the front element. */
    int peek() {
        if(s1.empty() && s2.empty()) //可以拿掉，题目条件:空队列不会调用pop peek
            throw runtime_error("Queue is empty!") ;
        if(s2.empty()){
            while(!s1.empty()){
                s2.push(s1.top());
                s1.pop() ;
            }
        }
        return s2.top() ;
    }

    /** Returns whether the queue is empty. */
    bool empty() {
        return s1.empty()&&s2.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: python-algorithm/ds_1_stack_queue/L232.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/2 10:08
# @Author  : Heng Li
# @File    : L232.py
# @Software: PyCharm

class MyQueue:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.stk1 = []
        self.stk2 = []

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.stk1.append(x)

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        if self.stk2 == []:
            while self.stk1:
                self.stk2.append(self.stk1.pop(-1))
        return self.stk2.pop(-1)

    def peek(self) -> int:
        """
        Get the front element.
        """
        if self.stk2 == []:
            while self.stk1:
                self.stk2.append(self.stk1.pop(-1))
        return self.stk2[-1]

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return self.stk1 == [] and self.stk2== []


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()